File: autoplay/drivers/bizhawk/driver.lua
-- autoplay BizHawk driver (DEV TOOL, WRITES INPUT, never shipped; agent_docs/phases/phase13.md, ADR 0071)
--
-- The piece inside BizHawk that carries out the autoplay core's commands. Load it through the dev
-- loader (dev-scripts/bizhawk-dev-loader.lua): add this file's absolute path to the instance's
-- control file. It picks its game module from AUTOPLAY_GAME (a global, or the environment) -- the
-- handoff names it -- and connects to the core on 127.0.0.1:AUTOPLAY_PORT (default 7870).
--
-- The link is the core's protocol 1, stated at the top of autoplay/driver/driver.go. One request
-- is carried out at a time; a press holds the controller for its frames and answers after them.
-- WHILE A PRESS IS RUNNING THIS SCRIPT HOLDS THE CONTROLLER. Take it off the target when done: an
-- input-driving tool left loaded is a suspect in every later report.

local PROTOCOL = 1
local MAX_LINE = 64 * 1024
-- Reconnect by the wall clock, not by frames: a connect attempt to a port nobody listens on blocks for
-- its timeout, and one every 30 frames took the emulator from 835 frames/s to 350 with no core running
-- (hookcost_probe.lua, 2026-09-16) -- at fast-forward, 30 frames is a few hundredths of a second.
local RETRY_SECONDS = 1
local REJECT_BACKOFF_SECONDS = 10
local PROGRAM_FRAME_LIMIT = 1800 -- a select across a long menu is far shorter; the core waits 40s

local function scriptDir()
	local info = debug.getinfo(1, "S")
	if info and info.source and info.source:sub(1, 1) == "@" then
		return info.source:sub(2):gsub("\\", "/"):match("^(.*)/[^/]*$")
	end
	return nil
end

local DIR = scriptDir()
if not DIR then
	error("autoplay driver: cannot find its own folder; load it by absolute path through the dev loader")
end
local ROOT = DIR:match("^(.*)/autoplay/drivers/bizhawk$")
if not ROOT then
	error("autoplay driver: expected to live at <repo>/autoplay/drivers/bizhawk, found " .. DIR)
end

local gameName = AUTOPLAY_GAME or os.getenv("AUTOPLAY_GAME")
local port = tonumber(AUTOPLAY_PORT or os.getenv("AUTOPLAY_PORT") or "") or 7870

-- One log per instance: two emulators writing one file cannot be told apart. The first instance, a game on
-- the default port, keeps the plain name.
local logName = port == 7870 and "driver_bizhawk.log"
	or string.format("driver_bizhawk_%s_%d.log", tostring(gameName):gsub("[^%w_]", "_"), port)
local logf = io.open(ROOT .. "/autoplay/runs/" .. logName, "a")
local function log(msg)
	local line = string.format("[%s f%d] %s", os.date("%H:%M:%S"), emu.framecount(), msg)
	if logf then
		logf:write(line, "\n")
		logf:flush() -- a handful of lines per command, never per frame
	else
		console.log("autoplay: " .. msg)
	end
end

local json = dofile(DIR .. "/json.lua")

-- LuaSocket, the copy the Emerald adapter vendors. lua54.dll first by full path, backslashes only:
-- the reasons are in meshghost_emerald.lua's "LuaSocket" block.
local LIB = (ROOT .. "/adapters/emulator/pokemon/emerald/lib/x64/"):gsub("/", "\\")
pcall(function() package.loadlib(LIB .. "lua54.dll", "autoplay_force_preload") end)
local openSocket, socketErr = package.loadlib(LIB .. "socket-windows-5-4.dll", "luaopen_socket_core")
if not openSocket then
	error("autoplay driver: could not load LuaSocket from " .. LIB .. ": " .. tostring(socketErr))
end
local socket = openSocket()

local game = nil
if gameName and gameName:match("^[%w_]+$") then
	game = dofile(DIR .. "/games/" .. gameName .. ".lua")
end

local sock, partial, state = nil, "", "down"
local nextTry = 0 -- os.time() at or after which to try connecting again
local queue = {}
local hold = nil
local lastDiff = nil

local function send(msg)
	if not sock then return false end
	local line = json.encode(msg)
	if #line + 1 > MAX_LINE then
		log("dropping an outgoing line of " .. #line .. " bytes")
		return false
	end
	local ok, err = sock:send(line .. "\n")
	if not ok then
		log("send failed: " .. tostring(err))
		return false
	end
	return true
end

local function close(why, backoff)
	if sock then pcall(function() sock:close() end) end
	sock, partial, state, queue, hold = nil, "", "down", {}, nil
	nextTry = os.time() + (backoff or RETRY_SECONDS)
	log("link down: " .. why)
end

local function reply(id, payload)
	send({ id = id, type = "result", payload = payload })
end

local function fail(id, message)
	send({ id = id, type = "error", payload = { message = message } })
end

local function changed(before, after)
	local a, b, out = game.diffKeys(before), game.diffKeys(after), {}
	for k, v in pairs(b) do
		if a[k] ~= v then out[k] = { from = a[k], to = v } end
	end
	return out
end

local function has(capability)
	for _, c in ipairs(game.capabilities) do
		if c == capability then return true end
	end
	return false
end

-- The answer to a press or a wait: how long, and what moved between before and after.
local function heldFor(frames)
	return function(after, _, _, before)
		return { frames = frames, before = before, after = after, changed = changed(before, after) }
	end
end

-- select: a program run one frame at a time over the module's observe().menu ({cursor, items, and
-- columns for a grid}) and its menuButtons ({prev, next, confirm, and left and right for a grid}). Every leg ends on the game's own state, never on a frame
-- count: press toward the entry until the menu's cursor changes, release for SETTLE frames, look
-- again; then hold confirm until the menu closes or changes. Returns a function, called once a frame,
-- that answers (pad or nil, finished, result or nil, error or nil) -- the shape of every program,
-- a game module's own included (game.programs).
local SETTLE, LEG_LIMIT = 2, 30

local function selectProgram(p)
	local keys = game.menuButtons
	local target, label, before, dir
	local phase, held, settle, steps, from, maxSteps = "look", 0, 0, 0, nil, 0

	local function sameMenu(m)
		if not m or m.window ~= before.window or #m.items ~= #before.items then return false end
		for i, item in ipairs(m.items) do
			if item ~= before.items[i] then return false end
		end
		return true
	end

	return function()
		-- game.menu() when the module has it: the menu alone, not a whole observation, every frame.
		local m
		if game.menu then m = game.menu() else m = game.observe().menu end
		if phase == "look" then
			before = m
			if type(m) ~= "table" or type(m.items) ~= "table" or #m.items == 0 then
				return nil, true, nil, "no menu is open (observe shows none)"
			end
			if p.index ~= nil then
				target = math.tointeger(p.index)
				if not target or target < 0 or target >= #m.items then
					return nil, true, nil, string.format("index %s is outside the menu's %d entries", tostring(p.index), #m.items)
				end
			else
				local want, hits = tostring(p.item):lower(), {}
				for i, item in ipairs(m.items) do
					if item:lower() == want then hits[#hits + 1] = i - 1 end
				end
				if #hits ~= 1 then
					return nil, true, nil, string.format("%d entries read %q; the menu has: %s", #hits, tostring(p.item), table.concat(m.items, " | "))
				end
				target = hits[1]
			end
			label, maxSteps, phase = m.items[target + 1], #m.items * 2 + 2, "move"
		end

		if settle > 0 then
			settle = settle - 1
			return nil, false
		end

		if phase == "move" then
			if not sameMenu(m) then return nil, true, nil, "the menu closed or changed while the cursor was moving" end
			if held == 0 and m.cursor == target then
				if p.confirm == false then
					return nil, true, { selected = label, index = target, steps = steps, confirmed = false }
				end
				phase = "confirm"
			else
				if held == 0 then
					if steps >= maxSteps then
						return nil, true, nil, string.format("the cursor is on %d after %d steps, not on %d", m.cursor, steps, target)
					end
					from, steps = m.cursor, steps + 1
					-- A menu with `columns` is a grid numbered row by row: reach the column first with
					-- the module's left/right, then the row with prev/next.
					local cols = math.tointeger(m.columns) or 1
					if cols > 1 and keys.left and keys.right and target % cols ~= m.cursor % cols then
						dir = (target % cols > m.cursor % cols) and keys.right or keys.left
					else
						dir = (target > m.cursor) and keys.next or keys.prev
					end
				elseif m.cursor ~= from then
					held, settle = 0, SETTLE
					return nil, false
				end
				held = held + 1
				if held > LEG_LIMIT then
					return nil, true, nil, string.format("the cursor did not move from %d on %s in %d frames", from, dir, LEG_LIMIT)
				end
				return { [dir] = true }, false
			end
		end

		-- confirm: hold until the menu is gone or no longer the same menu with the cursor on the entry.
		if held > 0 and (not sameMenu(m) or m.cursor ~= target) then
			return nil, true, { selected = label, index = target, steps = steps, confirmed = true }
		end
		held = held + 1
		if held > LEG_LIMIT then
			return nil, true, nil, string.format("the menu did not respond to %s in %d frames", keys.confirm, LEG_LIMIT)
		end
		return { [keys.confirm] = true }, false
	end
end

-- Start carrying out one request. Returns true when it answered at once.
local function begin(req)
	local verb, p = req.type, req.payload or {}
	local capability = verb
	if verb == "cheat" then capability = "cheat:" .. tostring(p.kind) end
	if not has(capability) then
		fail(req.id, "this driver does not support " .. tostring(capability))
		return true
	end
	if verb == "observe" then
		-- true: the agent asked, so the module may add what it leaves out of a press's before and after.
		reply(req.id, game.observe(true))
		return true
	elseif verb == "screenshot" then
		-- The game frame only (client.screenshot), never the window, into the game's own shots
		-- folder (the play-game skill's references/screenshots.md). A drawn-tier ghost is not in it.
		local name = type(p.name) == "string" and p.name or ""
		if not name:match("^[%w_%-]+$") or #name > 64 then
			fail(req.id, "screenshot needs a name of letters, digits, _ or -")
			return true
		end
		local path = string.format("%s/dev-scripts/shots/%s/autoplay_%s.png", ROOT, game.shots, name)
		local ok, err = pcall(function() client.screenshot(path) end)
		local fh = ok and io.open(path, "rb")
		if not fh then
			fail(req.id, "client.screenshot did not write " .. path .. ": " .. tostring(err))
			return true
		end
		fh:close()
		reply(req.id, { path = path, frame = emu.framecount() })
		return true
	elseif verb == "press" then
		local frames = math.tointeger(p.frames)
		if type(p.buttons) ~= "table" or #p.buttons == 0 or not frames or frames < 1 or frames > 600 then
			fail(req.id, "press needs buttons and 1 to 600 frames")
			return true
		end
		local known, pad = joypad.get(), {}
		for _, b in ipairs(p.buttons) do
			if type(b) ~= "string" or known[b] == nil then
				local names = {}
				for name in pairs(known) do names[#names + 1] = name end
				table.sort(names)
				fail(req.id, "unknown button " .. tostring(b) .. "; this core has " .. table.concat(names, ", "))
				return true
			end
			pad[b] = true
		end
		hold = { id = req.id, pad = pad, left = frames, before = game.observe(), finish = heldFor(frames) }
		log(string.format("press %s for %d frames", table.concat(p.buttons, "+"), frames))
		return false
	elseif verb == "wait" then
		-- Frames pass with NO input: never hold a button to wait, since every button means
		-- something somewhere (B backs out of a menu).
		local frames = math.tointeger(p.frames)
		if not frames or frames < 1 or frames > 3600 then
			fail(req.id, "wait needs 1 to 3600 frames")
			return true
		end
		hold = { id = req.id, pad = nil, left = frames, before = game.observe(), finish = heldFor(frames) }
		return false
	elseif verb == "select" then
		if type(game.menuButtons) ~= "table" then
			fail(req.id, "this game module names no menu buttons")
			return true
		end
		if (p.item == nil) == (p.index == nil) then
			fail(req.id, "select needs exactly one of item and index")
			return true
		end
		log("select " .. tostring(p.item or p.index))
		hold = { id = req.id, program = selectProgram(p), before = game.observe(), count = 0 }
		return false
	elseif verb == "snapshot" or verb == "restore" then
		-- The core names the file: a named state under its gitignored states folder, never a numbered
		-- slot, so nothing here can touch slot 1 or any rig's slot. BizHawk's savestate.save/load take a
		-- path (tasvideos.org/Bizhawk/LuaFunctions); the core checks the file itself afterwards.
		local path = p.path
		if type(path) ~= "string" or not path:match("%.State$") then
			fail(req.id, verb .. " needs a path ending in .State")
			return true
		end
		if verb == "snapshot" then
			local ok, err = pcall(function() savestate.save(path) end)
			if not ok then
				fail(req.id, "savestate.save failed: " .. tostring(err))
				return true
			end
			log("snapshot " .. path)
			reply(req.id, { path = path, frame = emu.framecount() })
			return true
		end
		local ok, loaded = pcall(function() return savestate.load(path) end)
		if not ok or loaded == false then
			fail(req.id, "savestate.load failed: " .. tostring(loaded))
			return true
		end
		log("restore " .. path)
		-- Whatever the module built up from hooks describes the memory before the load, not after.
		if game.restored then game.restored() end
		-- Answer one frame later, from the loaded state.
		hold = { id = req.id, left = 1, before = nil, finish = function(after) return { path = path, after = after } end }
		return false
	elseif verb == "cheat" then
		local run = game.cheats and game.cheats[p.kind]
		if not run then
			fail(req.id, "no cheat " .. tostring(p.kind) .. " in this game module")
			return true
		end
		local before = game.observe()
		local plan, err = run(type(p.args) == "table" and p.args or {})
		if not plan then
			fail(req.id, tostring(err))
			return true
		end
		log("cheat " .. tostring(p.kind))
		hold = {
			id = req.id, left = 0, before = before, untilFn = plan.untilFn, limit = plan.limit or 1, count = 0,
			finish = function(after, done, count)
				-- plan.report, when the module has one, reads the cheat's effect back from the game.
				return { kind = p.kind, done = done, frames = count, before = before, after = after, changed = changed(before, after),
					report = plan.report and plan.report() or nil }
			end,
		}
		return false
	end
	local make = game.programs and game.programs[verb]
	if make then
		-- A program may name its own frame limit (a long route outlasts a menu).
		local program, err, limit = make(p)
		if not program then
			fail(req.id, tostring(err))
			return true
		end
		log(verb)
		hold = { id = req.id, program = program, before = game.observe(), count = 0, limit = limit }
		return false
	end
	fail(req.id, "unhandled request " .. tostring(verb))
	return true
end

local function handle(line)
	local ok, msg = pcall(json.decode, line)
	if not ok or type(msg) ~= "table" then
		log("ignoring a line that does not parse")
		return
	end
	if msg.type == "welcome" then
		state = "ready"
		log("welcomed by the core on port " .. port)
	elseif msg.type == "reject" then
		local reason = type(msg.payload) == "table" and msg.payload.reason or "?"
		close("rejected: " .. tostring(reason), REJECT_BACKOFF_SECONDS)
	elseif msg.id and state == "ready" then
		queue[#queue + 1] = msg
	end
end

local function connect()
	local s = socket.tcp()
	if not s then return end
	s:settimeout(0.05)
	if not s:connect("127.0.0.1", port) then
		pcall(function() s:close() end)
		nextTry = os.time() + RETRY_SECONDS
		return
	end
	s:settimeout(0)
	pcall(function() s:setoption("tcp-nodelay", true) end)
	sock, partial, state = s, "", "hello_sent"
	send({
		type = "hello",
		payload = {
			protocol = PROTOCOL,
			host = "bizhawk",
			game = game.game,
			variant = game.variant,
			build = game.build(),
			capabilities = game.capabilities,
			protected_slots = game.protected_slots,
		},
	})
	log("connected to 127.0.0.1:" .. port .. ", hello sent")
end

local function drain()
	while sock do
		local line, err, part = sock:receive("*l", partial)
		if line then
			partial = ""
			handle(line)
		elseif err == "timeout" then
			partial = part or ""
			if #partial > MAX_LINE then close("a line over " .. MAX_LINE .. " bytes") end
			return
		else
			close(tostring(err))
			return
		end
	end
end

-- Every frame: each key of the game module's watch() that changed goes out as a "<key>_changed" event
-- (map_changed, mode_changed, and whatever else the module watches). Without a watch(), map and mode.
local function watchEvents()
	local d
	if game.watch then
		d = game.watch()
	else
		local k = game.diffKeys(game.observe())
		d = { map = k.map, mode = k.mode }
	end
	if lastDiff and state == "ready" then
		for key, value in pairs(d) do
			if value ~= lastDiff[key] then
				send({ type = "event", payload = { kind = key .. "_changed", from = lastDiff[key], to = value, frame = emu.framecount() } })
			end
		end
	end
	lastDiff = d
end

if not game then
	log("no game module: set AUTOPLAY_GAME (e.g. emerald) before loading; doing nothing")
else
	log(string.format("loaded for %s (%s), core port %d", game.game, game.variant, port))
	if game.start then log(game.start()) end
end

MESHGHOST_DEV_TICK = function()
	if not game then return end
	if not sock then
		if os.time() >= nextTry then connect() end
		return
	end
	drain()
	if not sock then return end

	if hold then
		if hold.program then
			-- One frame of a program: it looks at what it needs, then either finishes or sets this frame's input.
			hold.count = hold.count + 1
			local pad, finished, result, err = hold.program()
			if not finished and hold.count > (hold.limit or PROGRAM_FRAME_LIMIT) then
				finished, err = true, string.format("still running after %d frames", hold.limit or PROGRAM_FRAME_LIMIT)
			end
			if finished then
				if err then
					fail(hold.id, err)
				else
					local o = game.observe()
					result.frames, result.before, result.after = hold.count, hold.before, o
					result.changed = changed(hold.before, o)
					reply(hold.id, result)
				end
				hold = nil
			elseif pad then
				joypad.set(pad)
			end
		elseif hold.left > 0 then
			if hold.pad then joypad.set(hold.pad) end
			hold.left = hold.left - 1
		elseif hold.untilFn then
			-- A leg that ends on the game's own state, with a frame limit so it cannot run forever.
			local after = game.observe()
			hold.count = hold.count + 1
			local done = hold.untilFn(after)
			if done or hold.count >= hold.limit then
				reply(hold.id, hold.finish(after, done, hold.count, hold.before))
				hold = nil
			end
		else
			-- The last set applies to the frame after it, so the answer is read one frame later.
			local after = game.observe()
			reply(hold.id, hold.finish(after, true, 0, hold.before))
			hold = nil
		end
	elseif #queue > 0 then
		begin(table.remove(queue, 1))
	end
	watchEvents()
end

MESHGHOST_DEV_UNLOAD = function()
	if game and game.stop then game.stop() end
	close("unloaded")
	if logf then logf:close() end
	logf = nil
end

File: autoplay/driver/driver.go
// Package driver is the core's side of the driver link: a loopback TCP listener that accepts ONE
// driver at a time — the small piece inside an emulator or game that carries out commands — and
// turns its newline-delimited JSON into request/response calls and a buffer of events.
//
// It is game-blind: it never reads a payload, only routes it. What a driver can do is whatever
// its hello says, and every name in a payload (a mode, a map, an event kind) is opaque here.
//
// The wire (protocol 1), one JSON object per line, at most MaxLineBytes:
//
//	driver -> core  {"type":"hello","payload":Hello}                 first line, within HelloTimeout
//	core -> driver  {"type":"welcome","payload":{"protocol":1}}      or {"type":"reject","payload":{"reason":"..."}}, then close
//	core -> driver  {"id":N,"type":"<verb>","payload":{...}}         a request
//	driver -> core  {"id":N,"type":"result","payload":{...}}         or {"id":N,"type":"error","payload":{"message":"..."}}
//	driver -> core  {"type":"event","payload":{"kind":"...",...}}    unsolicited, buffered by the hub
//
// A hello's "persisting", and a cheat's answer carrying the same field, list the cheats still in effect
// in the game (a noclip left on): the only payload field the core reads, so a run segment begun while
// one is on is not labelled walked.
package driver

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net"
	"sync"
	"time"
)

// Protocol is the link version a driver must say in its hello.
const Protocol = 1

// MaxLineBytes caps one line in either direction, the same cap the MeshGhost bridge uses.
const MaxLineBytes = 64 * 1024

// HelloTimeout is how long a new connection has to send its hello.
const HelloTimeout = 5 * time.Second

// EventBuffer is how many of the newest events the hub keeps.
const EventBuffer = 512

// Hello is what a driver says about itself on connect.
type Hello struct {
	Protocol       int      `json:"protocol"`
	Host           string   `json:"host"`
	Game           string   `json:"game"`
	Variant        string   `json:"variant,omitempty"`
	Build          string   `json:"build,omitempty"`
	Capabilities   []string `json:"capabilities"`
	ProtectedSlots []int    `json:"protected_slots,omitempty"`
	// Persisting lists the cheat kinds in effect when the driver connected (package comment).
	Persisting Kinds `json:"persisting,omitempty"`
}

// Kinds is a list of cheat kinds as a driver sends it. A Lua driver's JSON cannot tell an empty array from an
// empty object, so {} reads as no kinds, the same as [].
type Kinds []string

// UnmarshalJSON takes an array of strings, or an empty object.
func (k *Kinds) UnmarshalJSON(b []byte) error {
	if len(b) > 0 && b[0] == '{' {
		var m map[string]json.RawMessage
		if err := json.Unmarshal(b, &m); err != nil {
			return err
		}
		if len(m) != 0 {
			return fmt.Errorf("a list of cheat kinds is an array, got an object with %d keys", len(m))
		}
		*k = nil
		return nil
	}
	var s []string
	if err := json.Unmarshal(b, &s); err != nil {
		return err
	}
	*k = s
	return nil
}

// Has reports whether the driver announced a capability.
func (h Hello) Has(capability string) bool {
	for _, c := range h.Capabilities {
		if c == capability {
			return true
		}
	}
	return false
}

// Event is one unsolicited driver message, numbered by the hub in arrival order.
type Event struct {
	Seq     uint64          `json:"seq"`
	At      time.Time       `json:"at"`
	Payload json.RawMessage `json:"payload"`
}

// ErrNoDriver is returned by Call when no driver is connected.
var ErrNoDriver = errors.New("no driver is connected")

// ErrDisconnected is returned by Call when the driver went away before answering.
var ErrDisconnected = errors.New("the driver disconnected before answering")

type envelope struct {
	ID      uint64          `json:"id,omitempty"`
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

type reply struct {
	payload json.RawMessage
	err     error
}

type conn struct {
	nc      net.Conn
	hello   Hello
	gen     uint64
	writeMu sync.Mutex
	done    chan struct{}
}

// Hub accepts driver connections on a loopback address and serves one at a time.
type Hub struct {
	ln  net.Listener
	log *log.Logger

	mu      sync.Mutex
	cur     *conn
	nextID  uint64
	pending map[uint64]chan reply
	events  []Event
	nextSeq uint64
	gen     uint64 // counts accepted drivers, so a caller can tell a reconnect from the same driver
}

// Listen binds addr, which must be a loopback address: a driver is a local process, and nothing
// off this machine may reach it.
func Listen(addr string, logger *log.Logger) (*Hub, error) {
	host, _, err := net.SplitHostPort(addr)
	if err != nil {
		return nil, fmt.Errorf("listen address %q: %w", addr, err)
	}
	ip := net.ParseIP(host)
	if ip == nil || !ip.IsLoopback() {
		return nil, fmt.Errorf("listen address %q is not a loopback IP; the driver link is local only", addr)
	}
	ln, err := net.Listen("tcp", addr)
	if err != nil {
		return nil, err
	}
	if logger == nil {
		logger = log.New(io.Discard, "", 0)
	}
	return &Hub{ln: ln, log: logger, pending: map[uint64]chan reply{}}, nil
}

// Addr is the address the hub is listening on.
func (h *Hub) Addr() net.Addr { return h.ln.Addr() }

// Close stops accepting and drops the current driver.
func (h *Hub) Close() error {
	err := h.ln.Close()
	h.mu.Lock()
	c := h.cur
	h.mu.Unlock()
	if c != nil {
		c.nc.Close()
	}
	return err
}

// Serve accepts drivers until the listener closes or ctx ends.
func (h *Hub) Serve(ctx context.Context) error {
	go func() {
		<-ctx.Done()
		h.Close()
	}()
	for {
		nc, err := h.ln.Accept()
		if err != nil {
			if ctx.Err() != nil || errors.Is(err, net.ErrClosed) {
				return nil
			}
			return err
		}
		go h.handle(nc)
	}
}

// Current returns the connected driver's hello.
func (h *Hub) Current() (Hello, bool) {
	hello, _, ok := h.CurrentConnection()
	return hello, ok
}

// CurrentConnection returns the connected driver's hello and its generation: every driver the hub
// accepts gets the next number, so the same number means the same connection.
func (h *Hub) CurrentConnection() (Hello, uint64, bool) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if h.cur == nil {
		return Hello{}, 0, false
	}
	return h.cur.hello, h.cur.gen, true
}

// EventsSince returns buffered events with Seq > since, oldest first, and the newest Seq seen.
func (h *Hub) EventsSince(since uint64) ([]Event, uint64) {
	h.mu.Lock()
	defer h.mu.Unlock()
	var out []Event
	for _, e := range h.events {
		if e.Seq > since {
			out = append(out, e)
		}
	}
	return out, h.nextSeq
}

// Call sends a request of type verb to the driver and waits for its answer, ctx's end, or the
// driver leaving. A driver's own error comes back as an error carrying its message.
func (h *Hub) Call(ctx context.Context, verb string, payload any) (json.RawMessage, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("encode %s payload: %w", verb, err)
	}
	h.mu.Lock()
	c := h.cur
	if c == nil {
		h.mu.Unlock()
		return nil, ErrNoDriver
	}
	h.nextID++
	id := h.nextID
	ch := make(chan reply, 1)
	h.pending[id] = ch
	h.mu.Unlock()

	defer func() {
		h.mu.Lock()
		delete(h.pending, id)
		h.mu.Unlock()
	}()

	if err := c.send(envelope{ID: id, Type: verb, Payload: body}); err != nil {
		return nil, fmt.Errorf("send %s: %w", verb, err)
	}
	select {
	case r := <-ch:
		return r.payload, r.err
	case <-c.done:
		return nil, ErrDisconnected
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

func (c *conn) send(e envelope) error {
	line, err := json.Marshal(e)
	if err != nil {
		return err
	}
	if len(line)+1 > MaxLineBytes {
		return fmt.Errorf("line of %d bytes is over the %d-byte cap", len(line)+1, MaxLineBytes)
	}
	c.writeMu.Lock()
	defer c.writeMu.Unlock()
	_, err = c.nc.Write(append(line, '\n'))
	return err
}

func (h *Hub) handle(nc net.Conn) {
	if tc, ok := nc.(*net.TCPConn); ok {
		// Small lines, one per command or event: never wait on Nagle.
		tc.SetNoDelay(true)
	}
	sc := bufio.NewScanner(nc)
	sc.Buffer(make([]byte, 0, 4096), MaxLineBytes)

	nc.SetReadDeadline(time.Now().Add(HelloTimeout))
	if !sc.Scan() {
		h.log.Printf("driver %s: no hello: %v", nc.RemoteAddr(), sc.Err())
		nc.Close()
		return
	}
	nc.SetReadDeadline(time.Time{})

	c := &conn{nc: nc, done: make(chan struct{})}
	var first envelope
	if err := json.Unmarshal(sc.Bytes(), &first); err != nil || first.Type != "hello" {
		h.reject(c, "the first line must be a hello")
		return
	}
	if err := json.Unmarshal(first.Payload, &c.hello); err != nil {
		h.reject(c, "the hello payload does not parse: "+err.Error())
		return
	}
	if c.hello.Protocol != Protocol {
		h.reject(c, fmt.Sprintf("protocol %d is not supported; this core speaks %d", c.hello.Protocol, Protocol))
		return
	}
	if c.hello.Host == "" || c.hello.Game == "" {
		h.reject(c, "a hello must name its host and game")
		return
	}

	h.mu.Lock()
	if h.cur != nil {
		h.mu.Unlock()
		h.reject(c, "busy: this core already has a driver")
		return
	}
	h.gen++
	c.gen = h.gen
	h.cur = c
	h.mu.Unlock()

	welcome, _ := json.Marshal(map[string]int{"protocol": Protocol})
	if err := c.send(envelope{Type: "welcome", Payload: welcome}); err != nil {
		h.drop(c, err)
		return
	}
	h.log.Printf("driver connected: host=%s game=%s variant=%s build=%s capabilities=%v",
		c.hello.Host, c.hello.Game, c.hello.Variant, c.hello.Build, c.hello.Capabilities)

	for sc.Scan() {
		var e envelope
		if err := json.Unmarshal(sc.Bytes(), &e); err != nil {
			h.log.Printf("driver: dropping a line that does not parse: %v", err)
			continue
		}
		switch e.Type {
		case "result", "error":
			h.deliver(e)
		case "event":
			h.record(e.Payload)
		default:
			h.log.Printf("driver: ignoring a line of type %q", e.Type)
		}
	}
	h.drop(c, sc.Err())
}

func (h *Hub) reject(c *conn, reason string) {
	body, _ := json.Marshal(map[string]string{"reason": reason})
	c.send(envelope{Type: "reject", Payload: body})
	h.log.Printf("driver %s rejected: %s", c.nc.RemoteAddr(), reason)
	c.nc.Close()
}

func (h *Hub) deliver(e envelope) {
	h.mu.Lock()
	ch, ok := h.pending[e.ID]
	h.mu.Unlock()
	if !ok {
		h.log.Printf("driver: an answer to id %d nobody is waiting for", e.ID)
		return
	}
	if e.Type == "error" {
		var msg struct {
			Message string `json:"message"`
		}
		json.Unmarshal(e.Payload, &msg)
		if msg.Message == "" {
			msg.Message = "the driver reported an error without a message"
		}
		ch <- reply{err: errors.New(msg.Message)}
		return
	}
	ch <- reply{payload: e.Payload}
}

func (h *Hub) record(payload json.RawMessage) {
	h.mu.Lock()
	defer h.mu.Unlock()
	h.nextSeq++
	h.events = append(h.events, Event{Seq: h.nextSeq, At: time.Now(), Payload: append(json.RawMessage(nil), payload...)})
	if len(h.events) > EventBuffer {
		h.events = append(h.events[:0], h.events[len(h.events)-EventBuffer:]...)
	}
}

func (h *Hub) drop(c *conn, err error) {
	c.nc.Close()
	h.mu.Lock()
	if h.cur == c {
		h.cur = nil
	}
	h.mu.Unlock()
	close(c.done)
	if err != nil {
		h.log.Printf("driver disconnected: %v", err)
	} else {
		h.log.Printf("driver disconnected")
	}
}
